src/context/typecore.ml

```diff
@@ -222,7 +222,31 @@ let add_local ctx k n t p =
 	ctx.locals <- PMap.add n v ctx.locals;
 	v
 
+let check_identifier_name ctx name kind p =
+	if starts_with name '$' then
+		display_error ctx (kind ^ " names starting with a dollar are not allowed") p
+	else
+		let is_valid_identifier s =
+			match s with
+			| "new" | "this" -> true
+			| _ -> Lexer.is_valid_identifier s
+		in
+		if not (is_valid_identifier name) then display_error ctx ("`" ^ name ^ "` is not a valid " ^ kind ^ " name") p
+
+let check_local_variable_name ctx name origin p =
+	let s_var_origin origin =
+		match origin with
+		| TVOLocalVariable -> "variable"
+		| TVOArgument -> "function argument"
+		| TVOForVariable -> "for variable"
+		| TVOPatternVariable -> "pattern variable"
```

Comment:
Hm, it just occurred to me that these are called "capture variables" in the "Capture variables must be lowercase" error, so should probably stay consistent with that?

Reply:
Or we rename the other one because capture variables is an overloaded term at the moment.

Reply:
Yeah, "pattern variable" does seem nicer, you've been confused more than once by calling these capture variables. :)

I wonder if the manual uses the term and would need to be adjusted too.

Reply:
There's a whole page on "variable capture", though "capture variable" technically only appears once at the bottom of the page: https://haxe.org/manual/lf-pattern-matching-variable-capture.html